Upload arweave function

lyfer233 opened the pull request:
Added:
1. Using the third package `arweave-rs` connect with the arweave.
2. Added a new field `arweave_id` to the kv_chains table and update the `src/schema.rs` file.
3. Updated two unit tests because of adding a field.
4. Add a unit test to ensure the connection with arweave and upload function.
5. Error process.

lyfer233 committed:
basic function finish

lyfer233 committed:
pass all unit test and correct some problems

lyfer233 committed:
Error process

nykma commented on migrations/2023_08_01-add_arewave_id_to_kv_chains/up.sql:
```diff
@@ -0,0 +1,5 @@
+-- Your SQL goes here
+
+ALTER TABLE kv_chains
+ADD arweave_id VARCHAR DEFAULT '';
```
Will `DEFAULT ''` be conflict with `Option<String>` ? I'm not sure about this.

nykma commented on src/controller/upload.rs:
```diff
@@ -47,8 +48,28 @@ pub async fn controller(request: Request) -> Result<Response, Error> {
 
+    // Try take the kvchain data upload to the arweave.
+    let arweave_document = KVChainArweaveDocument{
+        avatar: avatar.unwrap(),
```
Will `unwrap()` here be paniced?

nykma commented on src/controller/upload.rs:
```diff
@@ -47,8 +48,28 @@ pub async fn controller(request: Request) -> Result<Response, Error> {
+        previous_uuid: None,
+    };
+    
+    new_kv.arweave_id = match arweave_document.upload_to_arweave().await {
```
```suggestion
    new_kv.arweave_id = arweave_document.upload_to_arweave().ok();
```

nykma commented on src/error/mod.rs:
```diff
@@ -52,6 +58,8 @@ impl Error {
             Error::Base64Error(_) => StatusCode::BAD_REQUEST,
             Error::UuidParseError(_) => StatusCode::BAD_REQUEST,
+            Error::UrlParseError(_) => StatusCode::BAD_REQUEST,
+            Error::ArweaveError(_) => StatusCode::BAD_REQUEST,
```
Basiclly Arweave upload error is not user's fault, so it should be `5xx`, like `INTERNAL_SERVER_ERROR`

nykma commented on src/model/arweave/mod.rs:
```diff
@@ -0,0 +1,60 @@
+use crate::error::Error;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct KVChainArweaveDocument {
```
Why creating a new `Model` here?

lyfer233 committed:
1. fix four problems mentioned by PR
2. Update the config file(added the JWT path and arweave url)

lyfer233 commented:
Current problem:
- `arweave_id` not normally add into response when user post to `/v1/kv`, so I will think how to design and refactor some codes to address this issue.

nykma commented on src/controller/upload.rs:
```diff
@@ -47,12 +48,30 @@ pub async fn controller(request: Request) -> Result<Response, Error> {
+        previous_uuid: None,
+    };
+    
+    new_kv.arweave_id = arweave_document.upload_to_arweave().await.ok();
```
Please make this upload procedure a "background job". 
Uploading should not "block" this upload request.

lyfer233 commented:
Next plans:
1. fix this problem https://github.com/NextDotID/kv_server/pull/12#issuecomment-1705489400
2. Use thread to run upload job
3. Add more unit test to ensure error process and corner condition.

lyfer233 committed:
1. Adding a `arweave_id` fields in `kv` table because we will iterate kv model to generate the response.

2. Adding a `previous_arweave_id` fields in `KVChainArweaveDocument` model since it can help developers to quickly query the last link in the arweave.

3. Making more unit test and refactor some codes in `src/model/kv/tests.rs`

lyfer233 committed:
take upload task into background

lyfer233 committed:
delete a unit test

nykma commented on src/controller/upload.rs:
```diff
@@ -47,13 +48,38 @@ pub async fn controller(request: Request) -> Result<Response, Error> {
     // All done. Build response.
     let response = query_response(&mut conn, &persona.public_key)?;
+
+    new_kv.arweave_id = arweave_upload_future.await.unwrap();
```
It still blocks `json_response` below.

What I mean is, spawning a tokio background job and leave it alone.

nykma reviewed (changes requested):
Good enough. Only a small thing to tweak.

lyfer233 committed:
document update

lyfer233 committed:
background job finished

lyfer233 committed:
1. Add two unit test in upload task
2. update a unit test in store `arweave_id` into database
3. add more comment at `src/model/arweave` and `src/model/kv_chains`

nykma reviewed (commented):
Good for now. I'll fix some trivial stuff.

nykma reviewed (approved):
Good for now. I'll fix some trivial stuff.
